### PR TITLE
perf: optimise composer send behaviour

### PR DESCRIPTION
**Pull request by hurali97**
### Details

This fixes major things wrong with `ReportScreen` and it's children, `ReportActionCompose` and `ComposerWithSuggestions`. The highlight is the `personalDetails` prop from `onyx` as on high traffic accounts `personalDetails` becomes a huge object and is expensive to pass as a prop and have the component render/ re-render. So each time a message is sent, we are manipulating onyx for the  `report_`, `reportActions_` and `personalDetails` keys first with the optimistic data and then with the successful data. Most of the `ReportScreen` and it's children doesn't really need `personalDetails` as a prop from onyx as they are passing that prop to certain utility functions to calculate for example, icon. So this `personalDetails` prop can be removed and we can use directly from the utility as we have `Onyx.connect` already present in the utility files.

Apart from this, some props including `report` and `reportActions` have been simplified to become lightweight so that the can be passed easily down to the children. Previously, we were passing the whole object and it had properties that we don't need. Also, to prevent unnecessary re-rendering of `ReportActionCompose` the empty state is lifted and moved to `ReportActionCompose` alongside the `SendButton` and `EmojiButton`. Also, `sortedReportActions` aren't really needed to pass to `ReportActionsListItemRenderer` as it uses the prop to only evaluate whether it's a group or not. So we have passed a new prop `displayAsGroup` to the component directly instead of passing the whole actions array.

### Fixed Issues

$ https://github.com/Expensify/App/issues/26347
PROPOSAL: https://github.com/Expensify/App/issues/26347#issuecomment-1748791267

#### Testing Steps:

- Open a Report and focus the composer
- Start typing some letters and then hit send
- Once the message is sent, start typing again and then press send
- No message should be present in the composer and the behaviour should feel smooth

- [X] Verify that no errors appear in the JS console

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/47336142/f12e7031-3fd1-4576-ae48-9c3ba1d94f39

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/47336142/7a108c17-d5ef-4eb3-a904-b123875693dc

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/47336142/ec878cfc-f2ff-49f6-9f2d-19ff7d3b9b7d

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/47336142/43ab0dc9-373d-4226-ac92-fdd7cea21384

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/47336142/b6f81d25-d62f-4de5-bb76-9205ee99731a

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/47336142/983298cb-5e53-4921-b661-4bf58375f451

</details>

### CODE REVIEW

**Comment by hurali97** [comment]
Figuring out why Reassure tests are failing and will update the PR accordingly.

**Comment by hurali97** [comment]
@aimane-chnaif All green now !

**Review by aimane-chnaif — COMMENTED** [review]
After this refactor, https://github.com/Expensify/App/issues/26347 issue still remaining right?

**Comment by hurali97** [comment]
> After this refactor, #26347 issue still remaining right?

@aimane-chnaif No this will fix that issue because the refactor happened following that issue.

I will take a look at your comments shortly.

File: src/pages/home/ReportScreen.js
Comment (hurali97) [reply src/pages/home/ReportScreen.js:509]: I agree and I tested it on all of the devices using same and different accounts !!

**Comment by hurali97** [comment]
@aimane-chnaif Are you able to review the PR again? 🙂

**Comment by hurali97** [comment]
@aimane-chnaif Resolved conflicts !

**Comment by hurali97** [comment]
> > > After this refactor, #26347 issue still remaining right?
> > 
> > 
> > @aimane-chnaif No this will fix that issue because the refactor happened following that issue.
> 
> I am asking because after code review, this PR still doesn't fix the root cause of #26347. I tested on android emulator:
> 
>  Screen.Recording.2023-11-15.at.2.28.35.PM.mov


@aimane-chnaif Can we try to test this on a real world scenario? For example, test it on a device since there's no connected keyboard there and the reason we see it in emulator is because we can type really quickly directly from the connected keyboard which is not a real world scenario.

I tested it on android device and it works greatly. Let me know if you face the same on real android device.

File: src/components/AvatarWithDisplayName.js
Comment (hurali97) [reply src/components/AvatarWithDisplayName.js:96]: Yeah that's why I was surprised 😄 So I have supplied the appropriate params now, I think !

**Comment by mountiny** [comment]
running it although there is a bunch of conflicts https://github.com/Expensify/App/actions/runs/6894604269

**Comment by hurali97** [comment]
@aimane-chnaif Conflicts resolved.

However, on one of my mid-lower end device I could repro this again on Prod build downloaded by QR. On my high end device, it wasn't reproducible.

https://github.com/Expensify/App/assets/47336142/af8e5943-7f6f-4291-a6d8-1039c64df272

Although this PR still fixes a lot of other performance issues especially immediately toggling the Send button when user starts to type. I would still want to investigate further on my mid-lower end device to see why the compose is not getting empty.

Here's the catch, I am going on vacation for the next week and will be available after that, so I can continue work on this. Till then, we can merge this PR and have someone else continue the work or I can continue the work when I am back. Let me know what fits the best !

**Comment by mountiny** [comment]
I don't think this pr should have solved the composer sent issue so that is fine as long as it does not make it worse

@aimane-chnaif has other testing been good? Can we go ahead with a merge?

**Comment by mountiny** [comment]
@hurali97 could you please sync up with main to resolve conflicts? thank you!

**Comment by hurali97** [comment]
@aimane-chnaif conflicts resolved 👍

**Comment by hurali97** [comment]
@aimane-chnaif Both of the screenshots look the same to me. Can you elaborate more ? 🙂

**Comment by hurali97** [comment]
@aimane-chnaif It's fixed now. We are passing the wrong order of parameters to `getIcons` in main but it worked and wasn't caught. Anyways, I fixed it now with retaining correct order of parameters.

Screenshot 2023-11-30 at 12 21 24 PM

File: src/pages/ReportDetailsPage.js
Comment (hurali97) [reply src/pages/ReportDetailsPage.js:256]: Oh yeah, correct !!

File: tests/unit/CalendarPickerTest.js
Comment (hurali97) [reply tests/unit/CalendarPickerTest.js:80]: I see, will do that once it's landed on main 👍

**Comment by hurali97** [comment]
@aimane-chnaif Okay I will revert in those places where profile picture is not being updated 👍 

Can you tell me also, when I update profile photo on mobile, is it supposed to be changed in real-time on web also? Like as you have opened the two in the video?

**Comment by hurali97** [comment]
@aimane-chnaif So on main there's also no real-time update of profile details. Refer to my below video where I have to change the page in order for refresh to happen. 

However on my PR as you reported, I fixed the issue related to profile picture not updating on android. Also in the header on web, the profile picture wasn't being updated even after changing the page, so I have fixed that as well.

### main:

https://github.com/Expensify/App/assets/47336142/1dbbc338-7cdc-417b-ac6d-7b0c7852bda7

### This PR:


https://github.com/Expensify/App/assets/47336142/771f2ee0-9d23-4553-8784-b0463dd9a099

**Comment by hurali97** [comment]
@aimane-chnaif Thanks for explaining 👍 I will test this again !

**Comment by hurali97** [comment]
@aimane-chnaif Okay so it works for me on main but not in real-time per se but after 25-30 sec update is reflected on the second account but I guess that might be due to a certain reasons. 

However, as you can see in video this specific little icon doesn't update after 30 seconds or even if I switch chat, so this is probably a bug specific to this little icon, in main.


https://github.com/Expensify/App/assets/47336142/ff57b127-0b47-4a22-bfc0-f48db1f99fd3

I will test this now on my PR and fix the correspondence.

**Comment by hurali97** [comment]
@aimane-chnaif So I have fixed the `personalDetails` not being updated on my PR and also fixed the issue that's present on main which doesn't update the little icon in LHN.

On both main and this PR, update takes around +-30ms to reflect, probably because I am on debug or my internet connection.

https://github.com/Expensify/App/assets/47336142/70bf4e4a-3583-4987-9ba6-1ba10bf012d7

**Comment by hurali97** [comment]
@aimane-chnaif Sure thing, I will do that today !!

**Comment by hurali97** [comment]
@aimane-chnaif are you able to test again? I tried to test myself as much as I know about the app and I think it looks good now 👍

**Review by aimane-chnaif — COMMENTED** [review]
I keep testing all affected pages

File: src/components/AnonymousReportFooter.tsx
Comment (hurali97) [reply src/components/AnonymousReportFooter.tsx:38]: I can explain. Previously, we were passing the `policies` prop to `ReportFooter` and it has `ReportAnonymousFooter` as a child so it re-renders whenever `policies` update. But since I removed that altogether, it was not updating in real-time anymore. So I decided to use the dedicated `policy` in the nearest component possible.

File: src/components/AvatarWithDisplayName.tsx
Comment (hurali97) [reply src/components/AvatarWithDisplayName.tsx:27]: I am not sure of this, I followed this approach from the other parts of the codebase.

File: src/components/AttachmentModal.js
Comment (hurali97) [reply src/components/AttachmentModal.js:404]: This change is basically to memoize `AttachmentModal` and avoid re-rendering `ComposerWithSuggestions` when not required. Hence I have also moved `ComposerWithSuggestions` to not remain a children of `AttachmentModal`.

**Comment by mountiny** [comment]
@aimane-chnaif lets focus on this one please 🥺

File: src/pages/home/ReportScreen.js
Comment (hurali97) [reply src/pages/home/ReportScreen.js:166]: This `isLoadingInitialReportActions` is already present in main and I only changed the redundant `_.isEmpty` condition and moved it to a separate variable with memo `isEmptyChat`. I haven't introduced this `isLoadingInitialReportActions ` boolean and is something that main already contans.

Comment (hurali97) [reply src/pages/home/ReportScreen.js]: Oh that... it's resolved now, I think it might be a typo from me but it was fixed after pulling in the latest main and resolving the conflicts. Thanks for flagging !

**Comment by hurali97** [comment]
@aimane-chnaif done, thanks 🙂

**Comment by hurali97** [comment]
@aimane-chnaif any leads on the testing? 🙂

**Comment by mountiny** [comment]
@aimane-chnaif Can you resolve conflicts locally and finish the review here please? thank you!

**Comment by mountiny** [comment]
@allroundexperts Has approved the PR here https://github.com/Expensify/App/pull/32336 which is build atop of this one so I would really love to get this one going

**Comment by mountiny** [comment]
@hurali97 Another conflicts

@aimane-chnaif could you approve the PR once all is tested?

**Comment by hurali97** [comment]
Coming from [here](https://callstack-hq.slack.com/archives/C05LX9D6E07/p1703844985106289?thread_ts=1703679928.194659&cid=C05LX9D6E07). Waiting on to fix Reassure Tests.

In the mean time, I will keep resolving conflicts so that we don't have huge conflicts at the end, when Reassure tests are fixed and we are following up again on this PR. Thanks.

**Comment by mountiny** [comment]
Restarting the reassure tests since this PR got merged https://github.com/Expensify/App/pull/33831

**Comment by ArekChr** [comment]
@hurali97 please update the branch with the latest main to sync new perf-test gh action

**Comment by mountiny** [comment]
Yep, my bad, though this would be taking the action from main

**Comment by hurali97** [comment]
@aimane-chnaif Conflicts resolved !

**Comment by mountiny** [comment]
> sorry there's another conflict, while I almost completed testing

I dont think this matters so much, lets assume the conflict is simple, can you post the results of your testing now please?

**Comment by hurali97** [comment]
> BUG: [web] Copy + paste image to composer doesn't work
> 
> Screenshot 2024-01-10 at 12 20 54 PM


How can I reproduce this? By copying an image and pasting it in compose input?

**Comment by mountiny** [comment]
Thanks @aimane-chnaif

**Comment by hurali97** [comment]
@aimane-chnaif It's fixed now 👍 

https://github.com/Expensify/App/assets/47336142/fe8eaedf-19af-47d0-bf73-e6fd18398bf4

**Comment by hurali97** [comment]
@aimane-chnaif  can you verify again? I have tried to reproduce this on main and it's reproducible on main. I recall there was some change merged to main a week ago, where this functionality was removed from `onSubmitComment`. 

See [here](https://github.com/Expensify/App/blob/1f7a97b59e269f974bf3823bb079d89f28f7f995/src/pages/home/ReportScreen.js#L329C1-L338C7) there's no `scrollToOffset` anymore.

https://github.com/Expensify/App/assets/47336142/cc399a4f-9406-40bd-9310-6858a0c06e5e

File: src/styles/index.ts
Comment (mountiny) [inline src/styles/index.ts:1888]: This is against our guidelines 😬

Comment (hurali97) [reply src/styles/index.ts:1888]: > This is against our guidelines 😬

Oh I apologize. 

> what's the root cause?

Basically, I moved the `EmojiButton` and `SendButton` from `ReportActionCompose` to `ComposerWithSuggestions` and then it started to happen. The reason was to keep the re-rendering in one place only and don't re-render `ReportActionCompose`. It was basically concerning the empty state, which was previously defined in `ReportActionCompose`.

 Can you guys point me where can I look for design guidelines? OR anything relevant.

Comment (hurali97) [reply src/styles/index.ts:1888]: Thanks, it seems quite complicated to move `SendButton` and `EmojiButton` to platform specific files and then apply the style there. I will see if I can find a way where we have to make less changes to the codebase,

Comment (hurali97) [reply src/styles/index.ts]: Alright, it's fixed now. I opted for a simpler approach and that's to take out `SendButton` and `EmojiButton` outside of the `View` encapsulating `Composer` in `ComposeWithSuggestions`.

**Comment by mountiny** [comment]
@aimane-chnaif is going through the testing now

**Comment by hurali97** [comment]
@aimane-chnaif Have you completed your testing ? Are there anymore bugs ?

**Comment by hurali97** [comment]
> No more bugs

Great, I will proceed with fixing the Reply bug tomorrow morning 👍

**Comment by hurali97** [comment]
https://github.com/Expensify/App/assets/47336142/1f38b61c-0f12-409a-a5f2-1638813b94fd

@aimane-chnaif Reply bug is fixed 👍

File: src/pages/home/report/ReportActionsListItemRenderer.js
Comment (hurali97) [reply src/pages/home/report/ReportActionsListItemRenderer.js:79]: Thanks for flagging this. Done 👍

File: src/pages/home/report/ReportActionsListItemRenderer.js
Comment (hurali97) [reply src/pages/home/report/ReportActionsListItemRenderer.js:119]: Here's the catch. Right now each update to Onyx collection result in a new object instance creation which breaks the memoization itself. So to fix it better, we can just pick what we need and construct a memoized object which will only update when a property changes.

**Comment by hurali97** [comment]
@aimane-chnaif I don't understand the issue clearly. Looks like main branch is not showing the Chat's welcome message and this branch is showing it. This looks like the ideal behaviour and maybe something recently has been merged to main?

**Comment by hurali97** [comment]
https://github.com/Expensify/App/assets/47336142/1e4a2585-756b-49f3-9279-fb0465d9fb74

@aimane-chnaif fixed the issue and also double-checked any missing properties.

**Comment by hurali97** [comment]
> I did full regression test as much as what QA team does 😄 Hopefully above bugs are the last ones 🙏

Thanks, I appreciate your effort 🙂 . I will let you know once I fix these 👍

**Comment by hurali97** [comment]
@aimane-chnaif I have fixed the above issues except for the composer focus [one](https://github.com/Expensify/App/pull/30168#issuecomment-1899887260). Sometimes it focuses and sometimes not, so I am having hard time figuring it out. I will focus on this bug on Monday morning. 👍

**Comment by mountiny** [comment]
@hurali97 more conflicts

**Comment by hurali97** [comment]
@aimane-chnaif It's ready 👍

**Comment by mountiny** [comment]
https://github.com/Expensify/App/actions/runs/7656273034

**Comment by hurali97** [comment]
@aimane-chnaif What's the proceedings now? 

TS checks and Linting is failing after pulling in the latest main, so I can't fix it.

**Comment by hurali97** [comment]
@aimane-chnaif Thanks for doing a regression. I have fixed these issues now 👍

**Review by aimane-chnaif — APPROVED** [review]
I think we can now merge this as no bugs found at this moment from our side.
But still be on a look out for any regressions QA is reporting since this is big refactor.

**Review by mountiny — APPROVED** [review]
Let's go then, code looks good to me, this was a massive effort, thanks everyone!

**Review by mountiny — APPROVED** [review]
Let's go then, code looks good to me, this was a massive effort, thanks everyone!

**Comment by rlinoz** [comment]
I believe this PR caused this regression: https://github.com/Expensify/App/issues/35559

File: src/pages/home/report/ReportActionsListItemRenderer.js
Comment (tgolen) [inline src/pages/home/report/ReportActionsListItemRenderer.js:81]: Is this referenced somehere? I can't find it anywhere.

Comment (mountiny) [reply src/pages/home/report/ReportActionsListItemRenderer.js:81]: Ah that would be for threads, it would have to be used for that little element showing how many replies are in thread, who responded, right?

**Pull request merged**
